marlowe-spec-test/src/Marlowe/Spec/Core/Guarantees.hs

```diff
@@ -0,0 +1,324 @@
+  ( Gen,
+    frequency,
+    suchThat,
+  )
+import Semantics
+  ( TransactionOutput (..),
+    TransactionOutputRecord_ext (..),
+    Transaction_ext (..),
+    inputs,
+    isQuiescent,
+    maxTimeContract,
+    txOutPayments,
+    txOutWarnings,
+    txOutContract,
+    emptyState, txOutState
+  )
+import SemanticsTypes
+  ( Contract (..),
+    State_ext (..),
+  )
+import SingleInputTransactions (traceListToSingleInput)
+import Test.QuickCheck (cover)
+import Test.QuickCheck.Monadic (PropertyM, assert, monitor, pre, run, stop)
+import Test.QuickCheck.Property (counterexample)
+import Test.Tasty (TestTree, testGroup)
+import Timeout (isClosedAndEmpty)
+import TransactionBound (maxTransactionsInitialState)
+
+-- Property based tests correponding to theorems defined in Isabelle.
+tests :: InterpretJsonRequest -> TestTree
+tests i = testGroup "Guarantees"
+    [
+    -- TransactionBound.thy
+    playTrace_only_accepts_maxTransactionsInitialStateTest i
+    -- SingleInputTransactions.thy
+    , traceToSingleInputIsEquivalentTest i
+    -- QuiescentResults.thy
+    , computeTransactionIsQuiescentTest i
+    , playTraceIsQuiescentTest i
+    -- Timeout.thy
+    , timedOutTransaction_closes_contractTest i
+    -- CloseIsSafe.thy
+    , closeIsSafeTest i
+    ]
+
+-- TransactionBound.thy
+--
+-- lemma playTrace_only_accepts_maxTransactionsInitialState :
+--    "playTrace sl c l = TransactionOutput txOut ⟹
+--      length l ≤ maxTransactionsInitialState c"
+playTrace_only_accepts_maxTransactionsInitialStateTest :: InterpretJsonRequest -> TestTree
+playTrace_only_accepts_maxTransactionsInitialStateTest interpret = reproducibleProperty "lemma playTrace_only_accepts_maxTransactionsInitialState" do
+    context <- run $ generateT $ genContext interpret
+    (contract, State_ext _ _ _ (Arith.Int_of_integer startTime) _, transactions) <- run $ generate $
+        frequency
+          [ (5, genContractStateAndValidTransactions context)
+          , (20, genContractStateAndValidTransactions context `suchThat` \(_,_,l) -> not (null l))
+          , (75, genContractStateAndValidTransactions context `suchThat` \(_,_,l) -> length l > 1)
+          ]
+    let
+        req :: Request JSON.Value
+        req = PlayTrace startTime contract transactions
+    RequestResponse res <- run $ liftIO $ interpret req
+
+    case JSON.fromJSON res of
+      JSON.Success (TransactionOutput TransactionOutputRecord_ext {}) -> do
+        monitor
+          ( counterexample $
+              "Request: " ++ showAsJson req ++ "\n"
+                ++ "Expected maxTransactionsInitialState (" ++ show (maxTransactionsInitialState contract) ++ ")\n"
+                ++ "to be an upper bound for the number of transactions (" ++ show (length transactions) ++ ")")
+        stop $ cover 50.0 (length transactions > 1) "more than one transaction" $ toInteger (length transactions) <= integer_of_nat (maxTransactionsInitialState contract)
+      JSON.Success (TransactionError _ ) -> pre False
```

Comment:
This can obscure a potential error. Tests should be discarded at generation time, not after making the test if they failed